sharkfuser/include/fusilli/attributes/tensor_attributes.h

```diff
@@ -9,6 +9,85 @@
+// (in-memory) layout - contiguous (NCHW) or channels-last (NHWC). Not all
+// operators support the non-contiguous channels-last format though as this
+// is specific to the backend that implements it.
+//
+// On the other hand, there is no notion of strides in MLIR / IREE compiler.
+// It expects the tensors to always be contiguous (row-major) as specified by
+// its shape. It therefore assumes physical layout is always the same as the
+// logical layout.
+//
+// Fusilli sits between the high-level ML frameworks (PyTorch) and the low-level
+// MLIR compilers (IREE), so it needs to conform to the constraints in tensor
+// representations between the frontend and the backend while supporting
+// both memory formats (for performance reasons). With this in mind, we follow
+// the following convention for memory formats in Fusilli:
+//
+// A tensor in Fusilli is defined by its logical shape which is always in
+// channels-first (NCHW if 4D) format. The physical (in-memory) layout
+// is then determined by the strides.
+//
+//  * Tensor T1 of logical shape (n,c,h,w) and stride (c*h*w, h*w, w, 1) is NCHW
+//  in-memory (also called contiguous or channels-first)
+//
+//  * Tensor T2 of logical shape (n,c,h,w) and stride (h*w*c, 1, w*c, c) is NHWC
+//  in-memory (also called non-contiguous or channels-last)
+//
+// Any other stride configuration is considered invalid and will result in an
+// error during validation.
+//
+// Fusilli emits MLIR assembly for the `torch` dialect which is tied to PyTorch
+// semantics where ops always operate on channels-first (NCHW if 4D) tensors.
+// This works fine when inputs to a Fusilli node are channels-first, but we need
+// special handling for when they are in channels-last format. We introduce the
+// applicable transposes to convert tensors from channels-last (NHWC if 4D) to
+// channels-first (NCHW if 4D) at the inputs and vice versa at the outputs of
+// the Fusilli node. This ensures that the MLIR assembly generated by Fusilli
+// for that node is compatible with the `torch` dialect semantics at the op
+// boundaries. The transposes are expected to eventually be propagated and
+// optimized out by the IREE compiler. In addition to this, we also transpose
+// the shapes for channels-last tensors in the generated MLIR assembly to
+// satisfy the contiguity requirement (physical layout == logical layout) in
+// MLIR. This allows us to drop the strides in the MLIR assembly.
+//
+// For example, here's what the initial and subsequent IRs for a convolution on
+// a channels-last tensor (NHWC) would look like as emitted by Fusilli and after
+// IREE's transpose elimination passes:
+//
+//  Fusilli graph:
+//    x [NCHW + stride] -> conv [NHWC] -> y [NCHW + stride]
+//
+//  Torch:
+//    x [NHWC] -> T [NCHW] -> conv [NCHW] -> T' [NHWC] -> y [NHWC]
+//
+//  Linalg:
+//    x [NHWC] -> conv [NHWC] -> T [NCHW] -> T' [NHWC] -> y [NHWC]
+//
+//    x [NHWC] -> conv [NHWC] -> y [NHWC]
+//
```

Comment:
Great description! I might label this these steps with their place in the flow as well as their representation:

```suggestion
//  User input (Fusilli graph):
//    x [NCHW + stride] -> conv [NHWC] -> y [NCHW + stride]
//
//  Generated ASM (Torch dialect):
//    x [NHWC] -> T [NCHW] -> conv [NCHW] -> T' [NHWC] -> y [NHWC]
//
//  Transpose elimination Passes (Linalg dialect):
//    x [NHWC] -> conv [NHWC] -> T [NCHW] -> T' [NHWC] -> y [NHWC]
//
//    x [NHWC] -> conv [NHWC] -> y [NHWC]
//
```

Reply:
Done